src/server.c

```diff
@@ -4495,13 +4492,25 @@ void populateCommandLegacyRangeSpec(struct redisCommand *c) {
     if (c->key_specs_num == 0)
         return;
 
-    if (c->key_specs_num == 1 &&
-        c->key_specs[0].begin_search_type == KSPEC_BS_INDEX &&
-        c->key_specs[0].find_keys_type == KSPEC_FK_RANGE)
-    {
-        /* Quick win */
-        c->legacy_range_key_spec = c->key_specs[0];
-        return;
+    if (c->key_specs_num == 1) {
+        /* Commands SPUBLISH/SSUBSCRIBE/SUNSUBSCRIBE to support sharded channels have
+         * key specs defined for client to be able to handle parsing of channels
+         * using the information from `COMMAND` command.
+         *
+         * However, the sharded channels aren't actual keys and are treated differently
+         * from the keys. Hence, we want to avoid setting legacy range key spec for
+         * commands around sharded channels.
+         */
+        if (c->key_specs[0].flags & CMD_KEY_SHARD_CHANNEL) {
+            return;
+        }
```

Comment:
i don't think we wanna do that.
the whole point was that an innocent client, who doesn't know anything about the command, will be able to forward it to the right cluster node.
so it is likely that such an innocent (old) client, is using the lagacy spec)

Reply:
I'm slightly confused regarding the keyspecs interaction with shard channel. Let me sync up with @madolson again and update here.

Reply:
No, that was not the point. Innocent clients can continue to be inefficient and do multi-hops. If a client wanted to be smart, they can use the newer keyspecs to understand the topology. The legacy ranges already have gaps that clients have to work around. 

Most cluster clients don't use this information, and just hack around it: https://github.com/redis/redis-rb/blob/543beb26a98f16b898b8ff9730b756e1c6b87e56/lib/redis/cluster/command.rb#L50.

Reply:
ok, so here's a counter example: https://github.com/go-redis/redis/blob/e9a8bb4f86553aeab7fa13302f3487bb0eaf9e03/command.go#L62 (one that uses the legacy firstkey)

Reply:
and another one: https://github.com/redis/redis-py/blob/04b8d34e212723974b9b1f484fe7cd9e93f0e315/redis/commands/parser.py#L68

so by exposing this info in the legacy firstkey, we may be getting implicit support by old clients

Reply:
i admit i didn't review the code deeply, so i'm not sure what you mean.
i imagined that when we make these look like keys (even though they're not), the amount of special checks will be lower, not higher.
i.e. in most cases the check of what to do with the command and it's arguments will benefit by making these look like keys. and we'll need to add special checks in just one or two places.

e.g. maybe this one can be reverted: https://github.com/redis/redis/pull/8621#discussion_r773671102

Reply:
Amusingly the redis-py case won't actually work, since it'll find the pubsub flags and decide that it has "no-keys" since it is not one of the hard coded list of commands.

Reply:
Yeah, it's a tradeoff between a touchpoint here: #8621 (comment) and in getNodesForQuery() (where it's looking for keys for cluster stuff) and touch points in modules, ACL, and tracking (where it's looking for keys for key related stuff). With more code refactoring I think all of these can be simplified though by relying on the full keyspecs instead of the legacy one, so I don't want to use that as a sticking point for which decision we pick.

Reply:
@oranagra Hari and I talked it over, and we decided to go with your suggestion. I believe the other touch points should eventually be resolved when we move more completely to key specs and can read the flags. We'll also update the documentation to make it clear that for compatibility reasons sharded-channels are considered keys for `COMMAND INFO`, `COMMAND`, and `COMMAND GETKEYS`.

Reply:
ok.. please try, and feel free to revert if you feel it's doing more damage than good...
another alternative is to somehow exclude them in the legacy spec inside redis, but reflect them anyway in the output of COMMAND command, for the benefit of clients (excluding the ones that did explicit code to skip it, like reids-py did in the current version).
i feel that it's likely that in my approach there are more places that will implicitly work, and less that need explicit handling than the alternative.. but try it and see what you end up with.